src/Compilers/CSharp/Portable/Syntax/Syntax.xml

```diff
@@ -2042,6 +2045,14 @@
     </Field>
   </Node>
 
+  <Node Name="ExpressionColonSyntax" Base="CSharpSyntaxNode">
+    <Kind Name="ExpressionColon"/>
+    <Field Name="Expression" Type="ExpressionSyntax"/>
```

Comment:
I think I like the use of `Expression` to cover some future scenarios as well (we were thinking of indexing patterns like `{ [0]: var x }`).
But that subsumes the simple identifier case, so I wonder if we could remove the NameColon part of `SubpatternSyntax` and just emulate it for backcompat (returning null when it's not a simple identifier).
Tagging @CyrusNajmabadi for advice.

Reply:
i ran into somethign simialar when wanting to support `using X = any_type_not_just_name;`.  Here's how i implemented it in my PR:

https://github.com/dotnet/roslyn/pull/50167/files?file-filters%5B%5D=.cs#diff-95477edd7a2f9bf55ac156fbeb22d71f973b06dc20a7034b373e8c2e8341d6f0R19

Reply:
>if we could remove the NameColon part of SubpatternSyntax and just emulate it for backcompat

In order to do that, we could either (1) make this change on `NameColonSyntax` type itself.

```xml
 <Node Name="NameColonSyntax" Base="CSharpSyntaxNode">
    <Field Name="Expression" Type="ExpressionSyntax"/>
  </Node>
```
Or (2) keep `ExpressionColonSyntax` but that involves a common type with a covariant `Name` override.

Something like:
```xml
 <Node Name="BaseNameColonSyntax" Base="CSharpSyntaxNode">
    <Field Name="Name" Type="ExpressionSyntax"/> 
  </Node>
 <Node Name="NameColonSyntax" Base="BaseNameColonSyntax">
    <Field Name="Name" Type="IdentifierNameSyntax" Override="true"/>
  </Node>
 <Node Name="ExpressionColonSyntax" Base="BaseNameColonSyntax">
    <Field Name="Name" Type="ExpressionSyntax" Override="true"/>
  </Node>
```
Probably not an option because we avoid runtime features that are only available in core.

I am not comfortable with neither of these changes. (1) adds a cast for the numerous usages of `NameColonSyntax` everywhere and (2) is just too much, *if* feasible at all.

Reply:
It might be a bit tricky here though, because of the colon token :-/
One approach might be to introduce `ExpressionOrNameColonSyntax` deriving from `NameColonSyntax`.

Reply:
>One approach might be to introduce ExpressionOrNameColonSyntax deriving from NameColonSyntax.

Shouldn't that be the other way around? (option (2) above)

Reply:
@CyrusNajmabadi Do you have any comments on this?

Reply:
Names are already expressions. So I would just have ExpressionColon and have NameColon derive from that.

Reply:
>Names are already expressions. So I would just have ExpressionColon and have NameColon derive from that.

ExpressionColon and  NameColon are not expressions. It would help if you illustrate the tree you are suggesting.

Reply:
On phone, but here's what I'm thinking

Reply:
Add ExpressionColon. It has the shape you expect. Add NameColon. It derives from that. It's Name is strongly typed and returns the expression `as`ed to a NameSyntax. 

Then we update this type to expose both NameSyntax and ExpressionSyntax. Only one can be set at a time. Neither is a bridge to the other. Both are optional (though we can doc that one week always be non null)

Reply:
This means ask old code continues to work, and it produces the same old trees. 

New user guide will produce the new trees. 

Code to process these trees will likely always do:

`var x = p.NameColon ?? p.ExpressionColon`. 

This will work with determining the right type here.

Reply:
So you're thinking `ExpressionColon : NameColon`? Syntax nodes are either abstract or sealed so that probably need a base?

Reply:
Well, I had namecolon deriving from ExpressionColon. 

And you're right about the abstract/sealed issue. So we'd need a BaseExpressionColon that both of these derived from.

Reply:
Ok, I managed to actually make the base a little useful. I'll go with [this tree](https://sharplab.io/#v2:C4LgbgpgTglgZgTwDQBMQGoA+BnYUCuAxsAAQAqA9gNYQB2JA3gL4CwAUAIYBGuUHxJAAIAmEgFEAHgAcojVm2wQOAGwgohogJIo6pEOOmzm7dgHpTJAOZ1oHYGvbde/UlI5RgMFRpIAhDooAchwAthAAwhTKFPQM7CQJQgDMJE54LgYymUZWEMAA3CTyicnk1HQkkdG0lDSxuQVF7PKKKmokbh5eyj7BYVUxIP5BoRFRMYzxiYIptBAA7iTauiR9EIwNhUUJzeytquoi2QO0QwEQayeTbCUzJBSQULA62a8M1gXy8mYWIRy0+BU7E6nm8R0u41iU1KD2gz3WkiyiNkAF4AHyrUb5XZsEHdHzDC6jK5xG4JO7LWikNYkdGvAJLHRU7FsJhAA) which I think reflects what you had in mind.

Reply:
I'm not completely confident with the result, but just did it to see how it looks like.

Reply:
Thanks a lot for the experiment. I think the resulting syntax design works out. @CyrusNajmabadi do you agree?